### PR TITLE
Pod Ready++ API

### PR DESCRIPTION
**Pull request by freehan**
**What this PR does / why we need it**: 
Implements API changes for [pod ready++](https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md)

```release-note
NONE
```

/assign @MrHohn for review
/assign @thockin for approval

### CODE REVIEW

**Comment by freehan** [comment]
fyi @yujuhong

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -2474,6 +2480,12 @@ type PodSpec struct {
 	// configuration based on DNSPolicy.
 	// +optional
 	DNSConfig *PodDNSConfig
+	// If specified, all readiness gates will be evaluated for pod readiness.
+	// A pod is ready when all its containers are ready AND
+	// all conditions specified in the readiness gates have value equal to "True"
+	// More info: https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md
```
Comment (MrHohn) [inline pkg/apis/core/types.go]: Humm, we should probably not link to a github link in the API spec?

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -2474,6 +2480,12 @@ type PodSpec struct {
 	// configuration based on DNSPolicy.
 	// +optional
 	DNSConfig *PodDNSConfig
+	// If specified, all readiness gates will be evaluated for pod readiness.
+	// A pod is ready when all its containers are ready AND
+	// all conditions specified in the readiness gates have value equal to "True"
```
Comment (MrHohn) [inline pkg/apis/core/types.go]: It seems like we are using `PodConditionType` for `ReadinessGate`, and "True" is not one of the valid values?

File: pkg/apis/core/types.go [hunk 3]
```diff
@@ -2378,6 +2378,12 @@ const (
 	TolerationOpEqual  TolerationOperator = "Equal"
 )
 
+// ReadinessGate contains the reference to a pod condition
+type ReadinessGate struct {
+	// ConditionType refers to a condition in the pod's condition list with matching type.
+	ConditionType PodConditionType
```
Comment (MrHohn) [inline pkg/apis/core/types.go]: Or should this be a string?
```go
ConditionType string
```

File: staging/src/k8s.io/api/core/v1/types.go [hunk 4]
```diff
@@ -2759,6 +2764,13 @@ type PodSpec struct {
 	// configuration based on DNSPolicy.
 	// +optional
 	DNSConfig *PodDNSConfig `json:"dnsConfig,omitempty" protobuf:"bytes,26,opt,name=dnsConfig"`
+
+	// If specified, all readiness gates will be evaluated for pod readiness.
+	// A pod is ready when all its containers are ready AND
+	// all conditions specified in the readiness gates have value equal to "True"
+	// More info: https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md
+	// +optional
+	ReadinessGates []ReadinessGate `json:"readinessGates,omitempty" protobuf:"bytes,28,opt,name=readinessGates"`
```
Comment (MrHohn) [inline staging/src/k8s.io/api/core/v1/types.go]: Why 28 but not 27?

**Comment by MrHohn** [comment]
Curious how would some of the [constraints](https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md#constraints) ("ReadinessGates can only be specified at pod creation" and "No Update allowed on ReadinessGates") being enforced?

File: pkg/apis/core/types.go (hunk 1 above)
Comment (freehan) [reply pkg/apis/core/types.go]: I saw other fields has links though. https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L37

File: pkg/apis/core/types.go (hunk 2 above)
Comment (freehan) [reply pkg/apis/core/types.go]: I should say the status of corresponding condition should be "True"

File: pkg/apis/core/types.go (hunk 3 above)
Comment (freehan) [reply pkg/apis/core/types.go]: ReadinessGate refers to a PodCondition. And PodConditionType is the key of the condition list. Although PodConditionType is a string

File: staging/src/k8s.io/api/core/v1/types.go (hunk 4 above)
Comment (freehan) [reply staging/src/k8s.io/api/core/v1/types.go]: 27 is taken above...

**Comment by freehan** [comment]
> Curious how would some of the constraints ("ReadinessGates can only be specified at pod creation" and "No Update allowed on ReadinessGates") being enforced?

The fields that allow changes are whitelisted. No need to add anything in validation to enforce it.

File: pkg/apis/core/types.go (hunk 2 above)
Comment (MrHohn) [reply pkg/apis/core/types.go]: Ah got it, wasn't aware that condition can have status :)

File: staging/src/k8s.io/api/core/v1/types.go (hunk 4 above)
Comment (MrHohn) [reply staging/src/k8s.io/api/core/v1/types.go]: Haha good catch, surprised that is out of order.

**Review by MrHohn — APPROVED** [review]
/lgtm

File: pkg/apis/core/validation/validation_test.go [hunk 5]
```diff
@@ -5865,6 +5865,106 @@ func TestValidatePodDNSConfig(t *testing.T) {
...
+	errorCases := [][]core.ReadinessGate{
+		{
+			{
+				ConditionType: core.PodConditionType("invalid/condition/type"),
+			},
+		},
+	}
+	for _, tc := range errorCases {
+		if errs := validateReadinessGates(tc, field.NewPath("field")); len(errs) == 0 {
+			t.Errorf("expected failure for %v", tc)
```
Comment (MrHohn) [inline pkg/apis/core/validation/validation_test.go]: nit: might be more clear to have tc.desc and log it with error

**Comment by freehan** [comment]
kind/feature
priority/important-soon
sig/network

File: pkg/apis/core/types.go [hunk 6]
```diff
@@ -2378,6 +2378,12 @@ const (
 	TolerationOpEqual  TolerationOperator = "Equal"
 )
 
+// ReadinessGate contains the reference to a pod condition
+type ReadinessGate struct {
```
Comment (thockin) [inline pkg/apis/core/types.go]: PodReadinessGate ?

File: pkg/apis/core/validation/validation.go [hunk 7]
```diff
@@ -3493,6 +3508,21 @@ func ValidatePodStatusUpdate(newPod, oldPod *core.Pod) field.ErrorList {
...
 
+// validatePodConditions tests if the custom pod conditions are valid.
+func validatePodConditions(conditions []core.PodCondition, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	systemConditions := sets.NewString(string(core.PodScheduled), string(core.PodReady), string(core.PodInitialized))
+	for i, condition := range conditions {
+		if systemConditions.Has(string(condition.Type)) {
+			continue
+		}
+		for _, msg := range validation.IsQualifiedName(string(condition.Type)) {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go]: This doesn't actually check for a fully-qualified name - `foobar` is allowed as is `example.com/foobar`.   You maybe want  something different?

Comment (thockin) [reply pkg/apis/core/validation/validation.go]: also want a testcase, I think

File: pkg/apis/core/validation/validation_test.go (hunk 5 above)
Comment (thockin) [reply pkg/apis/core/validation/validation_test.go]: or at least `[%d]` with the testcase index

**Review by freehan — COMMENTED** [review]
Rebased.

**Comment by freehan** [comment]
`PodReadinessGate` feature gate default to false per https://github.com/kubernetes/kubernetes/pull/64344#issuecomment-393670239

**Comment by freehan** [comment]
Rebased

**Comment by jberkus** [comment]
Why is this release-note-none?

Doesn't an API change require a release note?

**Comment by bboreham** [comment]
Should this PR be listed under https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md#implementation-history ?

**Pull request merged**